src/Viewer/services/utils.js

```diff
@@ -44,27 +44,42 @@ const modifyPage = (action, currentPage, requestedPage, pages) => {
 };
 
 /**
- * Updates the URL based on the provided file metadata. If file
- * was downloaded using filePath, indicate the location in URL.
+ * Get modified URL from `window.location` based on the provided search and
+ * hash parameters.
  *
- * @param {object} fileMetadata
- * @param {number} logEventIdx
+ * @param {object} searchParams
+ * @param {object} hashParams
+ * @return {string} modified URL
  */
-const modifyFileMetadata = (fileMetadata, logEventIdx) => {
-    if (fileMetadata && fileMetadata.filePath) {
-        let url = `${window.location.origin}${window.location.pathname}`;
-        url += `?filePath=${fileMetadata.filePath}`;
-        if (0 !== logEventIdx) {
-            url += `#logEventIdx=${logEventIdx}`;
+const getModifiedUrl = (searchParams, hashParams) => {
+    const url = new URL(`${window.location.origin}${window.location.pathname}`);
+
+    const urlSearchParams = new URLSearchParams(window.location.search.substring(1));
+    const urlHashParams = new URLSearchParams(window.location.hash.substring(1));
+
+    Object.entries(searchParams).forEach(([key, value]) => {
+        urlSearchParams.delete(key);
+        if (null !== value) {
+            urlSearchParams.set(key, value.toString());
         }
-        window.history.pushState({path: url}, "", url);
-    } else if (fileMetadata && !fileMetadata.filePath) {
-        let url = `${window.location.origin}${window.location.pathname}`;
-        if (0 !== logEventIdx) {
-            url += `#logEventIdx=${logEventIdx}`;
+    });
+    Object.entries(hashParams).forEach(([key, value]) => {
+        urlHashParams.delete(key);
+        if (null !== value) {
+            urlHashParams.set(key, value.toString());
         }
-        window.history.pushState({path: url}, "", url);
+    });
+
+    let urlSearchParamsAsString = urlSearchParams.toString();
+    if (false === /%3F|%23|%26/.test(urlSearchParamsAsString)) {
```

Comment:
For now, only those 3 characters are identified to cause ambiguity for `new URLSearchParams()` to parse the `filePath` URL entirely. 

`/` is one that we confirmed not harmful (`URLSearchParams` still parses the entire `filePath` URL when `/` is inside the parameter).

I will experiment more with the other _Reserved Characters_ specified in https://datatracker.ietf.org/doc/html/rfc3986#section-2.2 and update here.

Reply:
We could also add a comment just stating the reasoning.

Reply:
In fact, I believe `%3F` (`?`) can be removed from the list. 

Below is an example to test what characters can cause `URLSearchParams` not to parse an entire parameter. 

```js
console.log("`#` would cause ambiguity")
let url = new URL("https://example.com/?shallEncode=!#$&'()*+,-/:;=?@[]");
let searchParams = new URLSearchParams(url.search);
for (const [key, value] of searchParams) {
    console.log(`${key}: ${value}`);
}
console.log("\n")

console.log("`&` would cause ambiguity")
url = new URL("https://example.com/?shallEncode=!$&'()*+,-/:;=?@[]");
searchParams = new URLSearchParams(url.search);
for (const [key, value] of searchParams) {
    console.log(`${key}: ${value}`);
}
console.log("\n")


console.log("The others should not cause ambiguity")
url = new URL("https://example.com/?shallEncode=!$'()*+,-/:;=?@[]");
searchParams = new URLSearchParams(url.search);
for (const [key, value] of searchParams) {
    console.log(`${key}: ${value}`);
}
console.log("\n")
```

Reply:
Gotcha